photon-ml/src/main/scala/com/linkedin/photon/ml/function/GeneralizedLinearModelLossFunction.scala

```diff
@@ -32,137 +32,109 @@ import org.apache.spark.rdd.RDD
-    */
-  override protected[ml] def hessianVector(
-      data: Iterable[LabeledPoint],
-      coefficients: Vector[Double],
-      multiplyVector: Vector[Double]): Vector[Double] = {
-
-    HessianVectorAggregator.calcHessianVector(
-      data, coefficients, multiplyVector, singleLossFunction, normalizationContext)
+    normalizationContext: ObjectProvider[NormalizationContext],
+    data: Either[RDD[LabeledPoint], Iterable[LabeledPoint]],
+    treeAggregateDepth: Int)
+  extends ObjectiveFunction(data, normalizationContext, treeAggregateDepth) with TwiceDiffFunction {
+
+  override protected[ml] def value(coefficients: Vector[Double]): Double =
+    calculate(coefficients)._1
+
+  override protected[ml] def gradient(coefficients: Vector[Double]): Vector[Double] =
+    calculate(coefficients)._2
+
+  override protected[ml] def calculate(coefficients: Vector[Double]): (Double, Vector[Double]) = data match {
+    case Left(dataAsRDD) =>
+      val broadcastCoefficients = dataAsRDD.sparkContext.broadcast(coefficients)
+      ValueAndGradientAggregator.calculateValueAndGradient(dataAsRDD,
+        broadcastCoefficients,
+        singleLossFunction,
+        normalizationContext,
+        treeAggregateDepth)
+
+    case Right(dataAsIterable) =>
+      ValueAndGradientAggregator.calculateValueAndGradient(dataAsIterable,
+        coefficients,
+        singleLossFunction,
+        normalizationContext)
   }
 
-  /**
-    * Compute the Hessian of the function under the given data and coefficients in the normalization context, then
-    * multiply it with a given multiplyVector.
-    *
-    * @param data The given data at which point to compute the hessian multiplied by a given vector
-    * @param broadcastedCoefficients The broadcasted model coefficients used to compute the hessian multiplied by a given
-    *                                vector
-    * @param multiplyVector The given multiplyVector to be multiplied with the Hessian. For example, in conjugate
-    *                       gradient method this multiplyVector would correspond to the gradient multiplyVector.
-    * @return The computed Hessian multiplied by the given multiplyVector
-    */
-  override protected[ml] def hessianVector(
-      data: RDD[LabeledPoint],
-      broadcastedCoefficients: Broadcast[Vector[Double]],
-      multiplyVector: Broadcast[Vector[Double]]): Vector[Double] = {
-
-    HessianVectorAggregator.calcHessianVector(
-      data, broadcastedCoefficients, multiplyVector, singleLossFunction, normalizationContext, treeAggregateDepth)
+  override protected[ml] def hessianVector(coefficients: Vector[Double], multiplyVector: Vector[Double])
+    : Vector[Double] = data match {
+    case Left(dataAsRDD) =>
+      val broadcastCoefficients = dataAsRDD.sparkContext.broadcast(coefficients)
+      val broadcastMultiply = dataAsRDD.sparkContext.broadcast(multiplyVector)
+      HessianVectorAggregator.calcHessianVector(
```

Comment:
Please read the FAQ of the original ```DiffFunction``` on whether we want to re-broadcast everything or not.

Or, please let me known the motivation of making such changes.